app/routes.py

```diff
@@ -1,2 +1,273 @@
+
+# modify incomplete task
+@task_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def make_task_incomplete(task_id):
+    task = Task.query.get(task_id)
+
+    if task is None:
+        return make_response(f"Task #{task_id} not found"), 404
+    else:
+        task.completed_at = None
+        db.session.commit()
+    response_body = {
+        "task": {
+            "id": task.task_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete":  False
+        }
+    }
+    return jsonify(response_body), 200
+
+
+# goals route
+@goals_bp.route("", methods = ["POST", "GET"], strict_slashes = False)
+def handle_goals():
+    if request.method == "POST":
+        request_body = request.get_json()
+        if "title" not in request_body:
+            return {"details": "Invalid data"}, 400
+        new_goal = Goal(title = request_body["title"])
+        db.session.add(new_goal)
+        db.session.commit()
+        response_body = {
+            "goal": new_goal.goal_to_json()
+        }
+        return jsonify(response_body), 201
+
+    #get all goals
+    if request.method == "GET":
+        goals = Goal.query.all()
+        response_body = []
+        for a_goal in goals:
+            response_body.append(a_goal.goal_to_json())
+        return jsonify(response_body), 200
+
+# get one goal
+@goals_bp.route("/<goal_id>", methods=["GET", "PUT", "DELETE"], strict_slashes = False)
+def handle_goal(goal_id):
+    goal = Goal.query.get(goal_id)
+
+    if request.method == "GET":
+        
+        if goal is None:
+            return make_response("", 404)
+
+        response_body = {
+            "goal": goal.goal_to_json()
+        }
+        return jsonify(response_body), 200
+
+    # Update a goal
+    elif request.method == "PUT":
+        if goal is None:
+            return make_response(f"Goal #{goal_id} not found"), 404
+        form_data = request.get_json()
+
+        if "title" not in form_data:
+            return {"details": "Invalid data"}, 400
+
+        goal.title = form_data["title"]
+        db.session.commit()
+
+        response_body = {
+            "goal": goal.goal_to_json()
+        }
+        return jsonify(response_body), 200
+
+    # Delete a goal
+    elif request.method == "DELETE":
+        if not goal:
+            return "", 404
+        db.session.delete(goal)
+        db.session.commit()
+
+        return {
+            "details": f"Goal {goal.goal_id} \"{goal.title}\" successfully deleted"
+        }
+# one-to-many relationship
+@goals_bp.route("/<goal_id>/tasks", methods = ["POST", "GET"], strict_slashes = False)
+def handle_task_to_goals(goal_id):
+    goal = Goal.query.get(goal_id)
+    if not goal:
+        return "", 404
+
+    if request.method == "POST":
+        request_body = request.get_json()
+        task_ids = request_body["task_ids"]
+
+        for task_id in task_ids:
+            goal_id.task_id = goal_id
```

Comment:
`goal_id` is  a simple string (of read directly from query params) or number, so it doesn't have a field called `task_id`. Even a `Goal` instance (like `goal`) wouldn't. A `Goal` has the field `tasks` (created by our relationship code) to which we can append `Task` instances (not IDs). On the other side of the relationship each `Task` instance has a field called `goal_id` that can be used to associate a `Task` with a `Goal`. With the changes suggested in `Goal`, they also have a field called `goal` to which we can assign an actual `Goal` instance.

For any of these cases, we really need to lookup each passed task id to get an instance, then either set the `goal`/`goal_id` on each task, or update the `tasks` collection on the goal. Something like

```python
            task = Task.query.get_or_404(task_id)
            task.goal = goal
```

This _wouldn't_ remove any existing tasks from the goal (we should think about what to do for those tasks).

Additionally, to ensure that the change to the relationships occur transactionally (either they all apply, or none of them do), we can move the `commit` outside the loop, so that it gets called once at the end.